### PR TITLE
many: add wait command and seeded target

### PR DESCRIPTION
**Pull request by mvo5**
This adds a `snap wait-seeded` command that is used by a new
`snapd.seeded` unit. This way other services can order after
snapd.seeded so that they can be sure that snapd is fully
working.

PoC for now (sprinting) - will add tests if the direction looks good.

I will also add (in a separate PR) the ability to add `after: [snapd]` or `after:[snapd.seeded]` which is needed by some classic snaps that use snapd and need to start after snapd.

### CODE REVIEW

**Comment by vorlonofportland** [comment]
Can you confirm that 'seeded', as used here, is a synchronization point that guarantees that all currently-installed snaps are mounted and available for use?  (Not just on first boot of an image with seeded snaps)

**Comment by chipaca** [comment]
@vorlonofportland no, snapd is not involved in the mounting-and-making-available process

File: data/systemd/snapd.seeded.service.in [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+Description=Wait until snapd is fully seeded
+After=snapd.service snapd.socket
+
+[Service]
+Type=oneshot
+ExecStart=@bindir@/snap wait-seeded
+RemainAfterExit=true
+
+[Install]
+WantedBy=multi-user.target
```
Comment (mvo5) [inline data/systemd/snapd.seeded.service.in]: Maybe we need to add `WantedBy=... cloud-final.target

**Comment by mvo5** [comment]
@niemeyer Happy to rename to `snapd.seed` if you prefer that. About the `snap watch` - I can wire it up there. I will need to handle the case when the seeding happend  a while ago and there is no change available in the state anymore that matches --last=seed. So it would be something like:
```
ExecStart=/bin/sh -c 'output=$(snap watch -type=seed) || if [ "$output" != "error: no changes found" ]; then echo "unexpected error: $output"; fi'
```
Alternatively we could add a new option to snap watch to not error if no change of the given type can be found.

**Comment by pedronis** [comment]
@mvo5 

> Alternatively we could add a new option to snap watch to not error if no change of the given type can be found.

those approaches are fragile because there's a little window at the start where snapd is not seeded nor there's a seed change yet!

**Comment by mvo5** [comment]
@pedronis Aha, very good point. So the extra command is the most robust option (?).

**Comment by pedronis** [comment]
> Aha, very good point. So the extra command is the most robust option (?).

yes, or we teach watch to special case seed (and later become-operational) to consider that they are initialisation changes meant to run at least once (this means looking for flag/state if the change is not there at all)

**Comment by niemeyer** [comment]
My main concern here is a sprawl in special case commands and APIs for edge cases. If feasible with a similar amount of effort, it would be better to consider what the problem is more generally.

That said, @pedronis has a good point here. The semantics around changes do not fit well with a concept that will need to be accounted for on every boot.

How about rethinking the issue in terms of configuration? Perhaps:
```
$ snap wait system seeded
```
Where _seeded_ is just a configuration variable on the core snap (now system). We can later expand the expressions supported by _wait_ to support richer queries, and can also improve support server side so we wait with the connection open. For the time being though, a simplistic approach that supports the simplest case with a client-side loop and presence of the variable alone might be as cheap as the current proposal.

How does that sound?

**Comment by pedronis** [comment]
@niemeyer that plan seems fine assuming that the general idea of letting snaps express some status/info through some (read-only for external when we have schemas)  subset of their config is something we are comfortable with/want to promote?

**Comment by niemeyer** [comment]
Tweaked the wording above to say that the conditional would actually check for the truthness of the value (true, number != 0, non-empty string, etc), instead of just presence, as it would be awkward to have a value that is _false_ yet evaluates as true for the condition.

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -371,6 +374,48 @@ func (m *DeviceManager) ensureBootOk() error {
 	return nil
 }
 
+func markSeededInConfig(st *state.State) error {
+	var seedDone bool
+	tr := config.NewTransaction(st)
+	if err := tr.Get("core", "seed.done", &seedDone); err != nil && !config.IsNoOption(err) {
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: why seed.done instead of seeded?

Comment (pedronis) [reply overlord/devicestate/devicemgr.go]: also that construction to be more readable would need to be "seeding.done"

Comment (pedronis) [reply overlord/devicestate/devicemgr.go]: also looking at "man systemd.special" most targets are based on names, not adjectives:

so snapd.seeding or snapd.seed

**Comment by mvo5** [comment]
The agreement (via irc) is `seed.loaded` for the config option and `<niemeyer> mvo: snapd.seeded.service`

File: cmd/snap/cmd_wait.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	addCommand("wait",
+		"Wait for configuration.",
+		"The wait command waits until a configration becomes true.",
+		func() flags.Commander {
+			return &cmdWait{}
+		}, nil, []argDesc{
+			{
+				name: "<snap>",
+				// TRANSLATORS: This should probably not start with a lowercase letter.
+				desc: i18n.G("The snap whose conf is being requested"),
```
Comment (niemeyer) [inline cmd/snap/cmd_wait.go]: "The snap for which configuration will be checked"

File: cmd/snap/cmd_wait.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+			},
+			{
```
Comment (niemeyer) [inline cmd/snap/cmd_wait.go]: }, {

File: cmd/snap/cmd_wait.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	return false
+}
+
+func trueish(vi interface{}) bool {
+	switch v := vi.(type) {
+	case bool:
+		if v == true {
+			return true
+		}
+	case int:
```
Comment (niemeyer) [inline cmd/snap/cmd_wait.go]: Missing int64, float32, float64?

File: daemon/response.go [hunk 6]
```diff
@@ -149,6 +149,8 @@ const (
 
 	errorKindNetworkTimeout      = errorKind("network-timeout")
 	errorKindInterfacesUnchanged = errorKind("interfaces-unchanged")
+
+	errorKindConfigNoSuchOption = errorKind("config-no-such-option")
```
Comment (niemeyer) [inline daemon/response.go]: "option-not-found" seems better here, per pattern established above (snap-not-found, app-not-found)

File: overlord/configstate/configcore/corecfg.go [hunk 7]
```diff
@@ -73,6 +73,7 @@ func Run(tr Conf) error {
 	if err := validateExperimentalSettings(tr); err != nil {
 		return err
 	}
+	// FIXME: ensure the user cannot set "core seed.loaded"
```
Comment (niemeyer) [inline overlord/configstate/configcore/corecfg.go]: Perhaps testing the change by just trying to tr.Get and making sure it hasn't changed before/after?

File: overlord/devicestate/devicemgr.go [hunk 8]
```diff
@@ -371,6 +374,48 @@ func (m *DeviceManager) ensureBootOk() error {
...
+	if !seedDone {
+		if err := tr.Set("core", "seed.loaded", true); err != nil {
+			return err
+		}
+		tr.Commit()
+	}
+	return nil
+}
+
+func (m *DeviceManager) ensureSeedInConfig() error {
```
Comment (niemeyer) [inline overlord/devicestate/devicemgr.go]: Do we need new method?  We already have a place changing "seeded" itself in state.. is that not an appropriate place for changing the configuration?

Comment (pedronis) [reply overlord/devicestate/devicemgr.go]: we also need to cover preexisting systems though,  so we need something in one of the ensure handlers for that

File: overlord/configstate/configcore/corecfg.go (hunk 7 above)
Comment (mvo5) [reply overlord/configstate/configcore/corecfg.go]: I need to look into this, we only get the "new" configuration from tr.Get() in configcore.Run() so to compare we need to get the state from tr.State(), get a new transaction with the current configuration and check if it is missing in the current config but is set in the new config and error only in this case.

**Comment by OddBloke** [comment]
@mvo5 What's the next step for getting this in to bionic? Is someone now preparing a release/SRU?

**Comment by mvo5** [comment]
@OddBloke We have some more pending PRs for 2.32 which are very close to land, once those are landed we do a 2.32.7 release and push that to bionic-proposed. Hopefully tomorrow.

File: overlord/devicestate/devicemgr.go (hunk 8 above)
Comment (mvo5) [reply overlord/devicestate/devicemgr.go]: Thanks @niemeyer - I pushed https://github.com/snapcore/snapd/pull/5157 that adds a better explanation and some more cleanups/tweaks for this PR (some more unit tests etc).

**Pull request merged**